Reusable TriggerModal component in RainbowKitCustomConnectButton and Faucet

0xChijoke opened the pull request:
## Description

Hello! This pull request introduces a reusable TriggerModal component and integrates it into the RainbowKitCustomConnectButton and Faucet components. This modification aims to improve developer/user experience and code reuseability.

https://github.com/scaffold-eth/scaffold-eth-2/assets/53488449/55b367aa-ae02-4c53-8ee6-3baef4c55f60


## Additional Information

- [x] I have read the [contributing docs](/scaffold-eth/scaffold-eth-2/blob/main/CONTRIBUTING.md) (if this is your first contribution)
- [x] This is not a duplicate of any [existing pull request](https://github.com/scaffold-eth/scaffold-eth-2/pulls)


Your ENS/address: chijoke.eth

rin-st commented:
Hi @0xChijoke , thank you for the pr!

I believe we should ask designers for big changes on the example-ui page.

So in this pr let's polish new modal component and use it for components where we have modals already - `RainbowKitCustomConnectButton` and `Faucet`.

We can add modal for example-ui page later if needed

rin-st commented on packages/nextjs/components/Modal.tsx:
```diff
@@ -0,0 +1,44 @@
+ */
+type ModalProps = {
+  trigger: JSX.Element;
+  modalTitle: string;
```
We need to use that prop for `htmlsFor` and `id` attributes but not for title. So I think `name` or `id' is better.

Title should be in different place, because it can be empty or long with spaces and can break labels/ids.

rin-st commented on packages/nextjs/components/Modal.tsx:
```diff
@@ -0,0 +1,44 @@
+type ModalProps = {
+  trigger: JSX.Element;
+  modalTitle: string;
+  modalContent: JSX.Element;
```
what about just using children instead of this prop

rin-st commented on packages/nextjs/components/Modal.tsx:
```diff
@@ -0,0 +1,44 @@
+  };
+};
+
+const Modal: React.FC<ModalProps> = ({ trigger, modalTitle, modalContent, action }) => {
```
Basically, it's not only modal, but also trigger. So maybe it should be named like `TriggerWithModal`? Check prop types name too if you'll change component name.

0xChijoke replied:
Sounds good.

Does TriggerModal work?

rin-st commented on packages/nextjs/components/Modal.tsx:
```diff
@@ -0,0 +1,44 @@
+  trigger: JSX.Element;
+  modalTitle: string;
+  modalContent: JSX.Element;
+  action?: {
```
It can be two or more buttons, for example "cancel" and "save". So probably we need to include this in content too

0xChijoke replied:
Exactly, the actions can be included in the modal content.

In the exampe-ui the send is in the modalContent.

The action is optional cause I had a specific use for passing it in.

It can also be removed if there is no need to leave the option there.

rin-st commented on packages/nextjs/components/example-ui/ContractInteraction.tsx:
```diff
@@ -19,6 +20,45 @@ export const ContractInteraction = () => {
     },
   });
 
+  // Define the modal trigger button
```
let's change this file when we'll have designs
https://github.com/scaffold-eth/scaffold-eth-2/issues/466

0xChijoke commented:
@rin-st  Let me know what you think about the current state of the component.

Should the optional action be removed because that can come with the children?

Would I need to create another PR to leave other files in there original state?

rin-st commented:
> Should the optional action be removed because that can come with the children?

Yes, we don't need it for now

> Would I need to create another PR to leave other files in there original state?

No, you can rename this pr and remove changes of other files. We can take those changes if needed from commits history.

When you complete the component, please change custom modals in `RainbowKitCustomConnectButton` and `Faucet`

Thanks!

rin-st commented on packages/nextjs/components/TriggerModal.tsx:
```diff
@@ -0,0 +1,42 @@
+  };
+};
+
+const Modal: React.FC<TriggerModalProps> = ({ trigger, name, children, action }) => {
```
name

rin-st commented on packages/nextjs/components/TriggerModal.tsx:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Modal component: A reusable modal component.
```
redundant comment

rin-st commented on packages/nextjs/components/TriggerModal.tsx:
```diff
@@ -0,0 +1,42 @@
+  trigger: JSX.Element;
+  name: string;
+  children: JSX.Element;
+  action?: {
```
remove for now

0xChijoke commented:
Hello @rin-st,

While refactoring the RainbowKitConnectButton component, I encountered scenarios where rendering the trigger separately from the modal body would be beneficial. 

To address this, I've split the component into two parts and introduced a new folder called TriggerWithModal. Within this folder, you'll find the Trigger and Modal components, each serving their specific roles and enhancing modularity.

I've already utilized this structure in both the RainbowKitConnectButton and Faucet components. Your feedback on this approach and any potential adjustments would be greatly appreciated.

Thanks!

rin-st commented:
Hi @0xChijoke , good job!

I don't like two-component nature of `TriggerWithModal` but it seems in this case it's a good solution.

I've tested your branch and found that modal doesn't close on click outside, as it was before. Fix it, please

rin-st commented on packages/nextjs/components/scaffold-eth/Faucet.tsx:
```diff
@@ -82,49 +84,42 @@ export const Faucet = () => {
   return (
     <div>
-      <label htmlFor="faucet-modal" className="btn btn-primary btn-sm px-2 rounded-full font-normal normal-case">
+      <Trigger id="faucet-modal" className="btn btn-primary btn-sm px-2 rounded-full font-normal normal-case">
```
and  it's better to move id to constant outside of component and use it in both trigger and modal

rin-st commented on packages/nextjs/components/scaffold-eth/RainbowKitCustomConnectButton.tsx:
```diff
@@ -134,10 +136,10 @@ export const RainbowKitCustomConnectButton = () => {
                       </li>
                       <li>
-                        <label htmlFor="qrcode-modal" className="btn-sm !rounded-xl flex gap-3 py-3">
+                        <Trigger id="qrcode-modal" className="btn-sm !rounded-xl flex gap-3 py-3">
```
same here

0xChijoke replied:
I've incorporated the changes you mentioned. 

I've also added an option for the modal to be closed on outside clicks(true by default), considering its reusability. 

This way, the modal can be used in various scenarios, allowing builders to choose the preferred behaviour.

Your patience and guidance are much appreciated. Looking forward to your feedback!

technophile-04 commented:
Hey guys thanks, but I don't see much advantages of adding this component : 
 
 With [DaisyUI v3 modal](https://daisyui.com/components/modal/#dialog-modal-with-a-close-button-at-corner): 

```jsx
<button onClick={() => window.faucet_modal.showModal()} className="...">
   ...
  <span>Faucet</span>
</button>

<dialog id={"faucet_modal"} className="modal">
  <form method="dialog" className="modal-box">
    ...
  </form>
</dialog>
```


With this component : 
```jsx
<Trigger id={FAUCET_MODAL_ID} className="...">
  ...
  <span>Faucet</span>
</Trigger>

<Modal id={FAUCET_MODAL_ID}>
  ...
</Modal>
```


Since we are using a component library (daisyUI v3) I think it's not worth creating our own micro components unless we have good reason to or there is something that is not present / not working in daisyUI

Also, I think by creating this component we are just abstracting 1-2 lines of code, I think its we let user use daisy UI and learn/customize as they want 

What do you guys think?

0xChijoke commented:
> Hey guys thanks, but I don't see much advantages of adding this component :
> 
> With [DaisyUI v3 modal](https://daisyui.com/components/modal/#dialog-modal-with-a-close-button-at-corner):
> 
> ```js
> <button onClick={() => window.faucet_modal.showModal()} className="...">
>    ...
>   <span>Faucet</span>
> </button>
> 
> <dialog id={"faucet_modal"} className="modal">
>   <form method="dialog" className="modal-box">
>     ...
>   </form>
> </dialog>
> ```
> 
> With this component :
> 
> ```js
> <Trigger id={FAUCET_MODAL_ID} className="...">
>   ...
>   <span>Faucet</span>
> </Trigger>
> 
> <Modal id={FAUCET_MODAL_ID}>
>   ...
> </Modal>
> ```
> 
> Since we are using a component library (daisyUI v3) I think it's not worth creating our own micro components unless we have good reason to or there is something that is not present / not working in daisyUI
> 
> Also, I think by creating this component we are just abstracting 1-2 lines of code, I think its we let user use daisy UI and learn/customize as they want
> 
> What do you guys think?



> 

I share your perspective. While the initial intention was to simplify modal usage, it seems that introducing custom components might actually lead to unnecessary complexity.

technophile-04 commented:
> I share your perspective. While the initial intention was to simplify modal usage, it seems that introducing custom components might actually lead to unnecessary complexity. 

Yeah, but tysm @0xChijoke for questing on this !! Also, Chijoke would really appreciate creating an issue or discussion before creating a new feature PR so that we can iterate and brainstorm on it if it is really required / others can put suggestions/improvements on it too 🙌

Really feel bad closing this PR but appreciate it !! Also, tysm to Rinat for reviewing this 🙌

rin-st commented:
> I think by creating this component we are just abstracting 1-2 lines of code, I think its we let user use daisy UI and learn/customize as they want

hmm, I'm not very agree, because
- why to force users to learn something if we can avoid it? yes it's a small thing, but it can take let's say 5-10 mins to configure it instead of two imports. Users can forget to add clickoutside and so on
- we have at least two modals for now, and if we want for example add shadow we need to do it in two places instead of one

Yes, if we want to add modal, we need to add most of the options from daisyui, and also add other components like `Button` and other simple things. It simplifies development but we need to think if we really need it.

Thanks @0xChijoke for your efforts and @technophile-04 for review!
